File: problems/tier2-intermediate/013-parking-lot/solution.js
'use strict';

/* Parking Lot — multi-floor + spot matching + pricing strategies. */

class ParkOp {
  constructor(kind, s1 = '', s2 = '', s3 = '', i1 = 0, i2 = 0, i3 = 0) {
    this.kind = kind;
    this.s1 = s1;
    this.s2 = s2;
    this.s3 = s3;
    this.i1 = i1;
    this.i2 = i2;
    this.i3 = i3;
  }
}

// Constants
const M_TYPE = 'MOTORCYCLE';
const C_TYPE = 'CAR';
const T_TYPE = 'TRUCK';

const SMALL = 0;
const MEDIUM = 1;
const LARGE = 2;

class Vehicle {
  constructor(licensePlate, type) {
    this.licensePlate = licensePlate;
    this.type = type;
  }
}

class ParkingSpot {
  constructor(spotId, floor, size) {
    this.spotId = spotId;
    this.floor = floor;
    this.size = size;
    this.isOccupied = false;
    this.vehicleLicensePlate = '';
  }
}

class Ticket {
  constructor() {
    this.ticketId = '';
    this.licensePlate = '';
    this.spotId = '';
    this.floor = 0;
    this.entryTime = 0;
    this.entryGateId = '';
    this.exitGateId = '';
  }
}

class Gate {
  constructor(gateId, type) {
    this.gateId = gateId;
    this.type = type;
  }
}

function _min_spot_size(vtype) {
  if (vtype === M_TYPE) return SMALL;
  if (vtype === C_TYPE) return MEDIUM;
  return LARGE;
}

function _is_compatible(spotSize, minRequired) {
  return spotSize >= minRequired;
}

class FlatRate {
  constructor(fee) {
    this.fee = fee;
  }
  calculateFee(durationSeconds) {
    return this.fee;
  }
}

class Hourly {
  constructor(rate) {
    this.rate = rate;
  }
  calculateFee(durationSeconds) {
    const hours = Math.ceil(durationSeconds / 3600.0);
    return this.rate * hours;
  }
}

class Tiered {
  constructor(base, mid, high) {
    this.base = base;
    this.mid = mid;
    this.high = high;
  }
  calculateFee(durationSeconds) {
    const hours = Math.ceil(durationSeconds / 3600.0);
    if (hours <= 1) return this.base;
    if (hours <= 3) return this.base + this.mid * (hours - 1);
    return this.base + this.mid * 2 + this.high * (hours - 3);
  }
}

class ParkingLot {
  constructor(numFloors) {
    this.floors = [];
    for (let i = 0; i < numFloors; i++) this.floors.push([]);
    this.activeTickets = new Map();
    this.gates = [];
    this.strategy = null;
    this.nextTicketId = 1;
  }

  addSpot(floor, size) {
    if (floor < 0 || floor >= this.floors.length) return;
    const spotId = 'F' + String(floor) + 'S' + String(this.floors[floor].length);
    this.floors[floor].push(new ParkingSpot(spotId, floor, size));
  }

  setPricingStrategy(s) {
    this.strategy = s;
  }

  addGate(gateId, type) {
    this.gates.push(new Gate(gateId, type));
  }

  getGates(type) {
    return this.gates.filter(g => g.type === type).map(g => g.gateId);
  }

  parkVehicle(vehicle, entryTime, gateId = '') {
    const minSize = _min_spot_size(vehicle.type);
    for (let f = 0; f < this.floors.length; f++) {
      const spots = this.floors[f];
      for (const spot of spots) {
        if (!spot.isOccupied && _is_compatible(spot.size, minSize)) {
          spot.isOccupied = true;
          spot.vehicleLicensePlate = vehicle.licensePlate;
          const tid = 'T' + String(this.nextTicketId);
          this.nextTicketId += 1;
          const t = new Ticket();
          t.ticketId = tid;
          t.licensePlate = vehicle.licensePlate;
          t.spotId = spot.spotId;
          t.floor = f;
          t.entryTime = entryTime;
          t.entryGateId = gateId;
          t.exitGateId = '';
          this.activeTickets.set(tid, t);
          return t;
        }
      }
    }
    return null;
  }

  unparkVehicle(ticketId, exitTime, gateId = '') {
    if (!this.activeTickets.has(ticketId)) return -1.0;
    const ticket = this.activeTickets.get(ticketId);
    ticket.exitGateId = gateId;
    for (const floorSpots of this.floors) {
      let broke = false;
      for (const spot of floorSpots) {
        if (spot.spotId === ticket.spotId && spot.isOccupied) {
          spot.isOccupied = false;
          spot.vehicleLicensePlate = '';
          broke = true;
          break;
        }
      }
      if (broke) {
        // Python only breaks the inner loop; keep outer loop continuing.
      }
    }
    const duration = exitTime - ticket.entryTime;
    let fee;
    if (this.strategy) {
      fee = this.strategy.calculateFee(duration);
    } else {
      fee = duration;
    }
    this.activeTickets.delete(ticketId);
    return fee;
  }

  getAvailableSpots(size) {
    let count = 0;
    for (const floorSpots of this.floors) {
      for (const spot of floorSpots) {
        if (!spot.isOccupied && spot.size === size) count += 1;
      }
    }
    return count;
  }

  getAvailableSpotsByFloor(floor, size) {
    if (floor < 0 || floor >= this.floors.length) return 0;
    let count = 0;
    for (const spot of this.floors[floor]) {
      if (!spot.isOccupied && spot.size === size) count += 1;
    }
    return count;
  }
}

function _size_from(s) {
  if (s === 'S' || s === 'small') return SMALL;
  if (s === 'M' || s === 'medium') return MEDIUM;
  return LARGE;
}

function _vtype_from(s) {
  if (s === 'M' || s === 'moto' || s === 'motorcycle') return M_TYPE;
  if (s === 'C' || s === 'car') return C_TYPE;
  return T_TYPE;
}

function _gate_from(s) {
  return s === 'entry' ? 'ENTRY' : 'EXIT';
}

function _fee_to_str(f) {
  if (f < 0) return '-1';
  return f.toFixed(2);
}

function parking_simulate(ops) {
  const out = [];
  let lot = null;
  let tickets = new Array(16).fill('');
  let snaps = [];
  const freshSnaps = () => {
    const arr = [];
    for (let i = 0; i < 16; i++) arr.push({ id: '', floor: -1, spotId: '', entryGate: '' });
    return arr;
  };
  snaps = freshSnaps();
  for (const op of ops) {
    const k = op.kind;
    if (k === 'new') {
      lot = new ParkingLot(op.i1);
      tickets = new Array(16).fill('');
      snaps = freshSnaps();
      out.push('ok');
    } else if (k === 'add_spot') {
      lot.addSpot(op.i1, _size_from(op.s1));
      out.push('ok');
    } else if (k === 'add_gate') {
      lot.addGate(op.s1, _gate_from(op.s2));
      out.push('ok');
    } else if (k === 'gates_count') {
      out.push(String(lot.getGates(_gate_from(op.s1)).length));
    } else if (k === 'gate_at') {
      const g = lot.getGates(_gate_from(op.s1));
      out.push(op.i1 >= 0 && op.i1 < g.length ? g[op.i1] : '');
    } else if (k === 'set_pricing') {
      let p = null;
      if (op.s1 === 'flat') p = new FlatRate(Number(op.i1));
      else if (op.s1 === 'hourly') p = new Hourly(Number(op.i1));
      else if (op.s1 === 'tiered') p = new Tiered(Number(op.i1), Number(op.i2), Number(op.i3));
      if (p) lot.setPricingStrategy(p);
      out.push('ok');
    } else if (k === 'park') {
      const v = new Vehicle(op.s1, _vtype_from(op.s2));
      const t = lot.parkVehicle(v, op.i1, op.s3);
      if (op.i2 >= 0 && op.i2 < tickets.length) {
        if (t) {
          tickets[op.i2] = t.ticketId;
          snaps[op.i2] = { id: t.ticketId, floor: t.floor, spotId: t.spotId, entryGate: t.entryGateId };
        } else {
          tickets[op.i2] = '';
          snaps[op.i2] = { id: '', floor: -1, spotId: '', entryGate: '' };
        }
      }
      out.push(t ? t.ticketId : '');
    } else if (k === 'ticket_at') {
      out.push(op.i1 >= 0 && op.i1 < tickets.length ? tickets[op.i1] : '');
    } else if (k === 'ticket_floor') {
      out.push(op.i1 >= 0 && op.i1 < snaps.length ? String(snaps[op.i1].floor) : '-1');
    } else if (k === 'ticket_spot_id') {
      out.push(op.i1 >= 0 && op.i1 < snaps.length ? snaps[op.i1].spotId : '');
    } else if (k === 'ticket_entry') {
      out.push(op.i1 >= 0 && op.i1 < snaps.length ? snaps[op.i1].entryGate : '');
    } else if (k === 'unpark') {
      const tid = op.i1 >= 0 && op.i1 < tickets.length ? tickets[op.i1] : '';
      const fee = lot.unparkVehicle(tid, op.i2, op.s1);
      out.push(_fee_to_str(fee));
    } else if (k === 'unpark_id') {
      const fee = lot.unparkVehicle(op.s1, op.i2, op.s2);
      out.push(_fee_to_str(fee));
    } else if (k === 'available') {
      out.push(String(lot.getAvailableSpots(_size_from(op.s1))));
    } else if (k === 'available_floor') {
      out.push(String(lot.getAvailableSpotsByFloor(op.i1, _size_from(op.s1))));
    } else {
      out.push('unknown:' + k);
    }
  }
  return out;
}

module.exports = { ParkOp, parking_simulate };

File: problems/tier2-intermediate/013-parking-lot/solution.py
"""Parking Lot — multi-floor + spot matching + pricing strategies."""

import math


class ParkOp:
    def __init__(self, kind, s1="", s2="", s3="", i1=0, i2=0, i3=0):
        self.kind = kind
        self.s1 = s1
        self.s2 = s2
        self.s3 = s3
        self.i1 = i1
        self.i2 = i2
        self.i3 = i3


# Constants
M_TYPE = "MOTORCYCLE"
C_TYPE = "CAR"
T_TYPE = "TRUCK"

SMALL = 0
MEDIUM = 1
LARGE = 2


class Vehicle:
    def __init__(self, licensePlate, type):
        self.licensePlate = licensePlate
        self.type = type


class ParkingSpot:
    def __init__(self, spotId, floor, size):
        self.spotId = spotId
        self.floor = floor
        self.size = size
        self.isOccupied = False
        self.vehicleLicensePlate = ""


class Ticket:
    def __init__(self):
        self.ticketId = ""
        self.licensePlate = ""
        self.spotId = ""
        self.floor = 0
        self.entryTime = 0
        self.entryGateId = ""
        self.exitGateId = ""


class Gate:
    def __init__(self, gateId, type):
        self.gateId = gateId
        self.type = type


def _min_spot_size(vtype):
    if vtype == M_TYPE:
        return SMALL
    if vtype == C_TYPE:
        return MEDIUM
    return LARGE


def _is_compatible(spot_size, min_required):
    return spot_size >= min_required


class FlatRate:
    def __init__(self, fee):
        self.fee = fee

    def calculateFee(self, durationSeconds):
        return self.fee


class Hourly:
    def __init__(self, rate):
        self.rate = rate

    def calculateFee(self, durationSeconds):
        hours = math.ceil(durationSeconds / 3600.0)
        return self.rate * hours


class Tiered:
    def __init__(self, base, mid, high):
        self.base = base
        self.mid = mid
        self.high = high

    def calculateFee(self, durationSeconds):
        hours = math.ceil(durationSeconds / 3600.0)
        if hours <= 1:
            return self.base
        if hours <= 3:
            return self.base + self.mid * (hours - 1)
        return self.base + self.mid * 2 + self.high * (hours - 3)


class ParkingLot:
    def __init__(self, numFloors):
        self.floors = [[] for _ in range(numFloors)]
        self.activeTickets = {}
        self.gates = []
        self.strategy = None
        self.nextTicketId = 1

    def addSpot(self, floor, size):
        if floor < 0 or floor >= len(self.floors):
            return
        spotId = "F" + str(floor) + "S" + str(len(self.floors[floor]))
        self.floors[floor].append(ParkingSpot(spotId, floor, size))

    def setPricingStrategy(self, s):
        self.strategy = s

    def addGate(self, gateId, type):
        self.gates.append(Gate(gateId, type))

    def getGates(self, type):
        return [g.gateId for g in self.gates if g.type == type]

    def parkVehicle(self, vehicle, entryTime, gateId=""):
        minSize = _min_spot_size(vehicle.type)
        for f, spots in enumerate(self.floors):
            for spot in spots:
                if not spot.isOccupied and _is_compatible(spot.size, minSize):
                    spot.isOccupied = True
                    spot.vehicleLicensePlate = vehicle.licensePlate
                    tid = "T" + str(self.nextTicketId)
                    self.nextTicketId += 1
                    t = Ticket()
                    t.ticketId = tid
                    t.licensePlate = vehicle.licensePlate
                    t.spotId = spot.spotId
                    t.floor = f
                    t.entryTime = entryTime
                    t.entryGateId = gateId
                    t.exitGateId = ""
                    self.activeTickets[tid] = t
                    return t
        return None

    def unparkVehicle(self, ticketId, exitTime, gateId=""):
        if ticketId not in self.activeTickets:
            return -1.0
        ticket = self.activeTickets[ticketId]
        ticket.exitGateId = gateId
        for floorSpots in self.floors:
            for spot in floorSpots:
                if spot.spotId == ticket.spotId and spot.isOccupied:
                    spot.isOccupied = False
                    spot.vehicleLicensePlate = ""
                    break
        duration = exitTime - ticket.entryTime
        if self.strategy:
            fee = self.strategy.calculateFee(duration)
        else:
            fee = float(duration)
        del self.activeTickets[ticketId]
        return fee

    def getAvailableSpots(self, size):
        count = 0
        for floorSpots in self.floors:
            for spot in floorSpots:
                if not spot.isOccupied and spot.size == size:
                    count += 1
        return count

    def getAvailableSpotsByFloor(self, floor, size):
        if floor < 0 or floor >= len(self.floors):
            return 0
        count = 0
        for spot in self.floors[floor]:
            if not spot.isOccupied and spot.size == size:
                count += 1
        return count


def _size_from(s):
    if s == "S" or s == "small":
        return SMALL
    if s == "M" or s == "medium":
        return MEDIUM
    return LARGE


def _vtype_from(s):
    if s == "M" or s == "moto" or s == "motorcycle":
        return M_TYPE
    if s == "C" or s == "car":
        return C_TYPE
    return T_TYPE


def _gate_from(s):
    return "ENTRY" if s == "entry" else "EXIT"


def _fee_to_str(f):
    if f < 0:
        return "-1"
    return f"{f:.2f}"


def parking_simulate(ops):
    out = []
    lot = None
    tickets = [""] * 16
    snaps = [{"id": "", "floor": -1, "spotId": "", "entryGate": ""} for _ in range(16)]
    for op in ops:
        k = op.kind
        if k == "new":
            lot = ParkingLot(op.i1)
            tickets = [""] * 16
            snaps = [{"id": "", "floor": -1, "spotId": "", "entryGate": ""} for _ in range(16)]
            out.append("ok")
        elif k == "add_spot":
            lot.addSpot(op.i1, _size_from(op.s1))
            out.append("ok")
        elif k == "add_gate":
            lot.addGate(op.s1, _gate_from(op.s2))
            out.append("ok")
        elif k == "gates_count":
            out.append(str(len(lot.getGates(_gate_from(op.s1)))))
        elif k == "gate_at":
            g = lot.getGates(_gate_from(op.s1))
            out.append(g[op.i1] if 0 <= op.i1 < len(g) else "")
        elif k == "set_pricing":
            p = None
            if op.s1 == "flat":
                p = FlatRate(float(op.i1))
            elif op.s1 == "hourly":
                p = Hourly(float(op.i1))
            elif op.s1 == "tiered":
                p = Tiered(float(op.i1), float(op.i2), float(op.i3))
            if p:
                lot.setPricingStrategy(p)
            out.append("ok")
        elif k == "park":
            v = Vehicle(op.s1, _vtype_from(op.s2))
            t = lot.parkVehicle(v, op.i1, op.s3)
            if 0 <= op.i2 < len(tickets):
                if t:
                    tickets[op.i2] = t.ticketId
                    snaps[op.i2] = {"id": t.ticketId, "floor": t.floor, "spotId": t.spotId, "entryGate": t.entryGateId}
                else:
                    tickets[op.i2] = ""
                    snaps[op.i2] = {"id": "", "floor": -1, "spotId": "", "entryGate": ""}
            out.append(t.ticketId if t else "")
        elif k == "ticket_at":
            out.append(tickets[op.i1] if 0 <= op.i1 < len(tickets) else "")
        elif k == "ticket_floor":
            out.append(str(snaps[op.i1]["floor"]) if 0 <= op.i1 < len(snaps) else "-1")
        elif k == "ticket_spot_id":
            out.append(snaps[op.i1]["spotId"] if 0 <= op.i1 < len(snaps) else "")
        elif k == "ticket_entry":
            out.append(snaps[op.i1]["entryGate"] if 0 <= op.i1 < len(snaps) else "")
        elif k == "unpark":
            tid = tickets[op.i1] if 0 <= op.i1 < len(tickets) else ""
            fee = lot.unparkVehicle(tid, op.i2, op.s1)
            out.append(_fee_to_str(fee))
        elif k == "unpark_id":
            fee = lot.unparkVehicle(op.s1, op.i2, op.s2)
            out.append(_fee_to_str(fee))
        elif k == "available":
            out.append(str(lot.getAvailableSpots(_size_from(op.s1))))
        elif k == "available_floor":
            out.append(str(lot.getAvailableSpotsByFloor(op.i1, _size_from(op.s1))))
        else:
            out.append("unknown:" + k)
    return out
